### PR TITLE
slider allows non-linear rate of change

### PR DESCRIPTION
**Pull request by billhails**
This adds an early/late slider that allows the rate of change of the tempo to vary. I've documented the maths and apart from that the code change is pretty simple.

The position of the slider controls the point at which the tempo is half-way between the start tempo and the final tempo, i.e. with a start tempo of 120bpm, an end tempo of 60bpm, a selection 4 ticks long and the slider at 1/4, the tempo will be 90bpm after 1 tick instead of 2.
If the slider is left in its default centre position, the behaviour should be identical with the current version.
The tempo changes remain smooth, e.g. it's a curve not two straight lines.

### CODE REVIEW

**Comment by jeetee** [comment]
Thanks for this, allow me a couple of weeks to review and possibly enhance it.
I think it would be useful for the user to have some visual feedback on his slider setting as well, and it can probably be limited to +-75%: if the user wanted an immediate change in Tempo, (s)he wouldn't be using this plugin.

File: TempoChanges.qml [hunk 1]
```diff
@@ -172,6 +172,50 @@ MuseScore {
...
+        //
+        //     1 = (2^p)(1 - l)
+        //
+        //     1/(1 - l) = 2^p
+        //
+        //     log(1/(1 - l)) = log(2^p) = p*log(2)
+        //
+        //     log(1/(1 - l))/log(2) = p
+        //
+        var power = Math.log(1 / (1 - linearity.value)) / Math.log(2);
```
Comment (jeetee) [inline TempoChanges.qml]: Shouldn't it be Math.log(1 / linearity.value) / Math.log(2); to not work reverse compared to how it does now?

Comment (billhails) [reply TempoChanges.qml]: I'm pretty sure it's correct, and I've manually tested both early and late for both rit. and accel. and saw reasonable-looking values in all cases, of course you should not take my word for that.

Re the 75% limit, could easily do, but then there will always be someone who wants a little more, maybe 95%?

Comment (billhails) [reply TempoChanges.qml]: Here's a link to a graph of the function for linearity values 1/16, 1/8, 1/4, 1/2, 3/4, 7/8 and 15/16:
https://www.desmos.com/calculator/wq8btekqba
you can see that for i.e. 1/4 (slider at 1/4) the curve cuts the y = 1/2 line at 0.25.
[edit] hmm, actually it doesn't, it cuts at 0.75, odd. Though I just re-checked and the plugin behaves as expected/described.

Comment (billhails) [reply TempoChanges.qml]: hold on this, there seems to be a mistake in my maths, the general behaviour is correct, but the precise values are wrong, i.e. slider at 1/4 doesn't map an input fraction of 1/4 to 1/2. I'll try to fix now.

**Comment by billhails** [comment]
I realised the maths wasn't quite correct, it was massively over-complicated too. I've just fixed that, tested locally and pushed a new version.
Here's a graph showing some sample curves: https://www.desmos.com/calculator/bo0uf8tirv

**Comment by Jojo-Schmitz** [comment]
Needs a rebase

**Comment by billhails** [comment]
Hi, sorry wasn't paying attention, done the rebase now.

**Comment by Jojo-Schmitz** [comment]
How does that "Merge pull request #10 from Jojo-Schmitz/translation" and "Add French translation" come into here? That doesn't look right. Neither should the "Merge branch 'master' of github.com:billhails/MuseScore_TempoChanges" be in this PR, nor "Update to ms3"

**Comment by billhails** [comment]
Hmm, Not sure, what I did was
```
git remote add upstream git@github.com:jeetee/MuseScore_TempoChanges.git
git fetch upstream
git rebase upstream/master
(fixed conflicts)
git add TempoChanges.qml
git rebase --continue
git push
```

**Comment by billhails** [comment]
Probably easiest is for me to just open a new PR, my change in itself is quite small, and close this one un-merged.

**Comment by Jojo-Schmitz** [comment]
I'm using `git pull --rebase upstream master` instead of `git rebase upstream/master` for things like this. Needs a `git push -f` then

**Comment by billhails** [comment]
I've created a second PR with just my changes, closing this one unmerged

**Pull request closed without merging**